include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h

```diff
@@ -400,30 +410,38 @@ struct __peer_prefix_helper<_OffsetT, __peer_prefix_algo::scan_then_broadcast>
     {
     }
 
-    ::std::uint32_t
-    __peer_contribution(_OffsetT& __new_offset_idx, _OffsetT __offset_prefix, bool __is_current_bucket)
+    template <typename _OffsetHistogramAcc>
+    _OffsetT
+    __peer_contribution(std::uint32_t __bucket, _OffsetHistogramAcc& __histogram)
     {
-        ::std::uint32_t __sg_item_offset = __dpl_sycl::__exclusive_scan_over_group(
-            __sgroup, static_cast<::std::uint32_t>(__is_current_bucket), __dpl_sycl::__plus<::std::uint32_t>());
-
-        __new_offset_idx |= __is_current_bucket * (__offset_prefix + __sg_item_offset);
-        // the last scanned value may not contain number of all copies, thus adding __is_current_bucket
-        ::std::uint32_t __sg_total_offset =
-            __dpl_sycl::__group_broadcast(__sgroup, __sg_item_offset + __is_current_bucket, __sg_size - 1);
-
-        return __sg_total_offset;
+        _OffsetT __offset = 0;
+        _ONEDPL_PRAGMA_UNROLL
+        for (::std::uint32_t __radix_state_idx = 0; __radix_state_idx < __radix_states; ++__radix_state_idx)
+        {
+            ::std::uint32_t __is_current_bucket = (__bucket == __radix_state_idx);
+            ::std::uint32_t __sg_item_offset = __dpl_sycl::__exclusive_scan_over_group(
+                __sgroup, static_cast<::std::uint32_t>(__is_current_bucket), __dpl_sycl::__plus<::std::uint32_t>());
+
+            __offset |= __is_current_bucket * (__histogram[__radix_state_idx] + __sg_item_offset);
+            // the last scanned value may not contain number of all copies, thus adding __is_current_bucket
+            ::std::uint32_t __sg_total_offset =
+                __dpl_sycl::__group_broadcast(__sgroup, __sg_item_offset + __is_current_bucket, __sg_size - 1);
+            __histogram[__radix_state_idx] += __sg_total_offset;
+        }
+        return __offset;
     }
 };
 
 #if _ONEDPL_SYCL_SUB_GROUP_MASK_PRESENT
-template <typename _OffsetT>
-struct __peer_prefix_helper<_OffsetT, __peer_prefix_algo::subgroup_ballot>
+template <std::uint32_t __radix_bits, typename _OffsetT>
+struct __peer_prefix_helper<__radix_bits, _OffsetT, __peer_prefix_algo::subgroup_ballot>
 {
     using _TempStorageT = __empty_peer_temp_storage;
 
     sycl::sub_group __sgroup;
     ::std::uint32_t __self_lidx;
     sycl::ext::oneapi::sub_group_mask __item_sg_mask;
+    static constexpr std::uint32_t __radix_states = 1 << __radix_bits;
```

Comment:
Instead of calculating `__radix_states` in each helper, maybe it makes sense to instead pass in `__radix_states` as a template parameter. It doesn't appear that `__radix_bits` is used for anything else besides calculating `__radix_states`.

Reply:
Done